sql-plugin/src/main/scala/org/apache/spark/sql/rapids/ExternalSource.scala

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import scala.util.{Failure, Success, Try}
+
+import com.nvidia.spark.rapids._
+
+import org.apache.spark.sql.avro.AvroFileFormat
+import org.apache.spark.sql.connector.read.Scan
+import org.apache.spark.sql.execution.FileSourceScanExec
+import org.apache.spark.sql.execution.datasources.FileFormat
+import org.apache.spark.sql.v2.avro.AvroScan
+import org.apache.spark.util.Utils
+
+object ExternalSource {
+
+  lazy val hasSparkAvroJar = {
+    val loader = Utils.getContextOrSparkClassLoader
+
+    /** spark-avro is an optional package for spark, so we should keep in mind that rapids can
+     * run successfully even without it */
+    Try(loader.loadClass("org.apache.spark.sql.v2.avro.AvroScan")) match {
+      case Failure(_) => false
+      case Success(_) => true
+    }
+  }
+
+  def tagSupportForGpuFileSourceScanExec(meta: SparkPlanMeta[FileSourceScanExec]): Unit = {
+    if (hasSparkAvroJar) {
+      meta.wrapped.relation.fileFormat match {
+        case _: AvroFileFormat => GpuReadAvroFileFormat.tagSupport(meta)
+        case f =>
+          meta.willNotWorkOnGpu(s"unsupported file format: ${f.getClass.getCanonicalName}")
+      }
+    } else {
+      meta.wrapped.relation.fileFormat match {
+        case f =>
+          meta.willNotWorkOnGpu(s"unsupported file format: ${f.getClass.getCanonicalName}")
+      }
+    }
+  }
+
+  def convertFileFormatForGpuFileSourceScanExec(format: FileFormat): FileFormat = {
+    if (hasSparkAvroJar) {
+      format match {
+        case _: AvroFileFormat => new GpuReadAvroFileFormat
+        case f =>
+          throw new IllegalArgumentException(s"${f.getClass.getCanonicalName} is not supported")
+      }
+    } else {
+      format match {
+        case f =>
+          throw new IllegalArgumentException(s"${f.getClass.getCanonicalName} is not supported")
+      }
+    }
```

Comment:
```suggestion
    format match {
      case _: AvroFileFormat if hasSparkAvroJar => new GpuReadAvroFileFormat
      case f =>
        throw new IllegalArgumentException(s"${f.getClass.getCanonicalName} is not supported")
    }
```

Reply:
I'd like to keep the original code just in case someone put another FileFormat behind `case _: AvroFileFormat ...` and cause the exception.

Reply:
But that's what the suggested code will do, it's semantically equivalent to what was written but simpler. Anything that isn't AvroFileFormat _and_ the avro jars are available will throw, just like the original code.  Or did you intend for the exception messages to be different?  As originally written, a match with a single, unqualified case is like writing
```
  if (true) {
     do something
  }
```
which of course can be simplified to just "do something".

Reply:
This case is a little special,

If we don't include spark-avro.jar (AvroFileFormat.class) in the classpath, then rapids plugin will throw exception

`java.lang.NoClassDefFoundError: org/apache/spark/sql/avro/AvroFileFormat`

so if dev later adds another format and put it behind 'case _: AvroFileFormat, then the exception will be thrown if no spark-avro jar